File: python/ngp.py
from dataclasses import dataclass
from enum import IntEnum
from typing import Iterable


class Scale(IntEnum):
    PERSON = 0
    INDIVIDUALS = 1
    FAMILY = 2
    COMMUNITY = 3
    NEIGHBORHOOD = 4
    INSTITUTION = 5
    CITY = 6
    COUNTY = 7
    REGION = 8
    STATE = 9
    COUNTRY = 10
    CONTINENT = 11
    PLANET = 12


SCALE_NAMES = [
    "person",
    "individuals",
    "family",
    "community",
    "neighborhood",
    "institution",
    "city",
    "county",
    "region",
    "state",
    "country",
    "continent",
    "planet",
]


@dataclass(frozen=True)
class ScaleInput:
    stakes_density: float = 0.0
    harmony: float = 0.0
    admissible: bool = True
    weight: float = 1.0


@dataclass(frozen=True)
class ScaleOutput:
    scale: Scale
    contribution: float
    constraint_failed: bool


@dataclass(frozen=True)
class Evaluation:
    score: float
    conflict_count: int
    constraint_failures: int
    outputs: tuple[ScaleOutput, ...]


def empty_inputs() -> list[ScaleInput]:
    return [ScaleInput() for _ in SCALE_NAMES]


def scale_name(scale: Scale) -> str:
    return SCALE_NAMES[int(scale)]


def evaluate(
    inputs: Iterable[ScaleInput],
    start: Scale = Scale.PERSON,
    end: Scale = Scale.PLANET,
) -> Evaluation:
    values = list(inputs)
    if len(values) != len(SCALE_NAMES):
        raise ValueError(f"expected {len(SCALE_NAMES)} scale inputs")

    start_index = int(start)
    end_index = int(end)
    if start_index > end_index:
        start_index, end_index = end_index, start_index

    score = 0.0
    conflict_count = 0
    constraint_failures = 0
    outputs: list[ScaleOutput] = []

    for index, value in enumerate(values):
        if index < start_index or index > end_index:
            outputs.append(ScaleOutput(Scale(index), 0.0, False))
            continue

        admissible = 1.0 if value.admissible else 0.0
        contribution = (
            value.stakes_density * value.harmony * admissible * value.weight
        )
        score += contribution

        if contribution < 0.0:
            conflict_count += 1
        if not value.admissible:
            constraint_failures += 1

        outputs.append(ScaleOutput(Scale(index), contribution, not value.admissible))

    return Evaluation(
        score=score,
        conflict_count=conflict_count,
        constraint_failures=constraint_failures,
        outputs=tuple(outputs),
    )

File: c/ngp.c
#include "ngp.h"

static const char *NSGP_SCALE_NAMES[NSGP_SCALE_COUNT] = {
    "person",
    "peers",
    "family",
    "community",
    "neighborhood",
    "institution",
    "city",
    "county",
    "region",
    "state",
    "country",
    "continent",
    "planet"
};

const char *nsgp_scale_name(NsgpScale scale)
{
    if (scale < 0 || scale >= NSGP_SCALE_COUNT) {
        return "unknown";
    }

    return NSGP_SCALE_NAMES[scale];
}

NsgpScaleInput nsgp_default_input(void)
{
    NsgpScaleInput input;
    input.stakes_density = 0.0;
    input.harmony = 0.0;
    input.admissible = 1;
    input.weight = 1.0;
    return input;
}

NsgpEvaluation nsgp_evaluate(
    const NsgpScaleInput inputs[NSGP_SCALE_COUNT],
    NsgpScale start,
    NsgpScale end
)
{
    NsgpEvaluation evaluation;
    size_t i;

    evaluation.score = 0.0;
    evaluation.conflict_count = 0;
    evaluation.constraint_failures = 0;

    for (i = 0; i < NSGP_SCALE_COUNT; i++) {
        evaluation.outputs[i].scale = (NsgpScale)i;
        evaluation.outputs[i].contribution = 0.0;
        evaluation.outputs[i].constraint_failed = 0;
    }

    if (start > end) {
        NsgpScale tmp = start;
        start = end;
        end = tmp;
    }

    for (i = (size_t)start; i <= (size_t)end && i < NSGP_SCALE_COUNT; i++) {
        const NsgpScaleInput input = inputs[i];
        const int admissible = input.admissible ? 1 : 0;
        const double contribution =
            input.stakes_density * input.harmony * (double)admissible * input.weight;

        evaluation.outputs[i].contribution = contribution;
        evaluation.outputs[i].constraint_failed = admissible ? 0 : 1;
        evaluation.score += contribution;

        if (!admissible) {
            evaluation.constraint_failures++;
        }

        if (contribution < 0.0) {
            evaluation.conflict_count++;
        }
    }

    return evaluation;
}
